### ISSUE TITLE
An infinity loop in the method java.util.concurrent.ConcurrentMap#computeIfPresent

### ISSUE THREAD

**Issue by QIvan** [hazelcast/hazelcast#11816]
Hi! This code have an infinity loop
 
```java

import com.hazelcast.config.Config;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.io.Serializable;
import java.math.BigDecimal;
import java.util.concurrent.ConcurrentHashMap;


public class Starter {

    public static final String ACCOUNTS = "accounts";

    public static void main(String[] args) {
        Config cfg = new Config();                                            // 1  
        HazelcastInstance instance = Hazelcast.newHazelcastInstance(cfg);     // 2
        IMap<Long, Account> mapCustomers = instance.getMap(ACCOUNTS);         // 3 
//        Map<Long, Account> mapCustomers = new ConcurrentHashMap<>();        // 4

        System.out.println("init");
        mapCustomers.put(1L, new Account(1L, "my"));
        System.out.println("compute");
        mapCustomers.computeIfPresent(1L, (id, account) -> {
            account.setValue(BigDecimal.TEN);
            return account;
        });
        System.out.println("done");

        mapCustomers.values().forEach(System.out::println);
    }

    @Data
    @EqualsAndHashCode
    @ToString
    public static class Account implements Serializable {
        private final long id;
        private final String name;
        private BigDecimal value = BigDecimal.ZERO;
    }

}
```
a output will be like this
```
init
compute
```

But if we comment lines 1-3 and uncomment line 4 the output will be 
```
init
compute
done
Starter.Account(id=1, name=my, value=10)
```
as expected. 

1. Hazelcast version: **3.9**
2. Cluster size: any 
3. Number of the clients: any 
4. Version of Java: 
```bash
$ java -version 
openjdk version "1.8.0_111"
OpenJDK Runtime Environment (build 1.8.0_111-b16)
OpenJDK 64-Bit Server VM (build 25.111-b16, mixed mode)
```
5. Operating system: any 
6. Logs and stack traces: -
7. Detailed description of the steps to reproduce your issue: run the code =) 
8,9 - no available

**Comment by QIvan** [comment]
There is this behavior because in `com.hazelcast.map.impl.operation.ReplaceIfSameOperation#run` we try to replace with a new value and before check that no modifications was between to apply the lambda and an attempt to replace. But the lambda has modified reference already and this check always will be failed because in `com.hazelcast.map.impl.recordstore.DefaultRecordStore#replace(com.hazelcast.nio.serialization.Data, java.lang.Object, java.lang.Object)` we compare the value by a key in the local map with a record in a cluster and there value by the same key has not modified yet.

**Comment by QIvan** [comment]
I know that a Hazelcast's target is Java 6, but in  Java 8 this is a bug. The method `computeIfAbsent` has javadoc 
```javadoc
     * If the value for the specified key is present and non-null, attempts to
     * compute a new mapping given the key and its current mapped value.
```

**Comment by taburet** [comment]
@QIvan Unfortunately, the issue can't be fixed at this moment. The root cause of the problem is that Hazelcast maps expect values to have value type semantics, while standard Java implementations expect reference type semantics. Hazelcast is still at Java 6, so we can't override `computeIfPresent` to provide our own implementation to fix this.

I have enhanced `IMap` javadoc to make things more clear about value types vs reference types and created a separate issue #11958 to provide our own `computeIfPresent implementation in Hazelcast 4.0, which will be based on a more recent Java version.

If you think there is more to this issue, feel free to reopen. If you are experiencing other problems, please open a new issue.

**Comment by mmedenjak** [comment]
Fixed in 4.1.
